---
title: "Whole genome structural variation analysis using Nanopore long DNA sequence reads"
date: "Report created: `r Sys.Date()`"
output:
  html_document:
    keep_md: yes
    number_sections: yes
    self_contained: yes
    theme: default
    highlight: null
    css: Static/ont_tutorial.css
    toc: yes
    toc_depth: 2
    toc_float:
      collapsed: yes
      smooth_scroll: yes
    df_print: paged
link-citations: yes
bibliography: Static/Bibliography.bib
always_allow_html: yes
---

<div style="position:absolute;top:0px;right:0px;padding:15px;background-color:gray;width:45%;">
```{r, echo=FALSE}
knitr::include_graphics("https://nanoporetech.com/themes/custom/nanopore/images/ont-logo.svg?tutorial=cas9enrichment")
```
</div>


```{r setup, include=FALSE}
knitr::opts_chunk$set(tidy=TRUE, tidy.opts=list(width.cutoff=80), 
                      fig.width=9, fig.height=6, warning=FALSE, message=FALSE, 
                      kable.force.latex=TRUE,
                      highlight=TRUE,
                      cache.path="Analysis/Cache/",
                      knitr.table.format = "html",
                      width = 120,
                      knitr.kable.NA = '')

options(knitr.kable.NA = '')

library(yaml)
library(session)
library(kableExtra)
library(scales) # required for comma
library(dplyr)
library(nanopoRe) # devtools::install_github("sagrudd/nanopoRe") # unloadNamespace("nanopoRe")
library(emojifont)
library(vcfR)
library(VariantAnnotation)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

config <- yaml.load_file("config.yaml")
#tutorialText   <- config$tutorialText
tutorialText <- FALSE

input_fastq <- file.path(basename(dirname(config$input_fastq)), basename(config$input_fastq))
reference_fasta <- file.path(basename(dirname(config$reference_fasta)), basename(config$reference_fasta))
sample_name <- config$sample_name
setRpath(file.path("Analysis", "R"))

### here lie dragons ###
# expeditions below this comment at your own risk
# welcome feedback and suggestions ...

slurpContent <- function(filename) {
  include = as.logical(tutorialText)
  if (include) {
    paste(readLines(filename),collapse="\n")
  }
}

setReferenceGenome(reference_fasta)
bamFile <- file.path("Analysis", sample_name, "alignment", paste0(sample_name, "_minimap2.bam"))
```


`r slurpContent("Static/TutorialPreamble.md")`




# Study design

The **`ont_tutorial_sv`** script has been used to generate this report. The configuration file, **`config.yaml`** has been parsed and the fastq format DNA sequence file has been identified as **` `r input_fastq` `**.

```{r fastqSummary, cache=FALSE, echo=FALSE}
if (!file.exists(input_fastq)) {
  cat(paste0("Fatal error - file ["+input_fastq+"] not found\n\n"))
  knitr::knit_exit()
}
validationResponse = fastqCheckup(input_fastq)
if (is.logical(validationResponse)) {
  if (validationResponse==FALSE) {
    cat(paste0("Fatal error - file ["+input_fastq+"] corrupt or not FASTQ format\n\n"))
    knitr::knit_exit()
  }
}

file.data <- t(data.frame(validationResponse[1:5]))

kable(file.data, format="html", caption="Table summarising provided FASTQ file content", booktabs=TRUE, table.envir='table*', linesep="", escape = FALSE, row.names=FALSE)  %>%
  kable_styling(c("striped", "condensed")) %>%
  footnote(symbol=c("Md5 checksum calculated using R digest package"), symbol_title="please note: ", footnote_as_chunk = TRUE)
```


# Mapping Summary

```{r mappingExecutiveSummary, echo=FALSE}
bamSummary <- bamSummarise(bamFile)
primaryBAM <- bamSummary[which(bamSummary$readFlag=="Primary"),]

infographicFile <- infoGraphicPlot4(identifier="mappingSummary", 
                 panelA=c(key="Primary Alignments", value=scales::comma_format()(nrow(primaryBAM)), icon="fa-sliders"), 
                 panelB=c(key="Mapping Yield", value=gdata::humanReadable(sum(primaryBAM$end - primaryBAM$start), standard="SI"), icon="fa-calculator"), 
                 panelC=c(key="Average Accuracy", value=paste0(round(mean(primaryBAM$accuracy)*100,digits=1),"%"), icon="fa-area-chart"), 
                 panelD=c(key="Average Identity", value=paste0(round(mean(primaryBAM$identity)*100,digits=1),"%"), icon="fa-info"))

 knitr::include_graphics(infographicFile)
```


# Mapping characteristics

```{r loadMapData, echo=FALSE, warning=FALSE, message=FALSE}



if (!file.exists(bamFile)) {
  cat(paste0("Fatal error - file ["+bamFile+"] not found\n\n"))
  knitr::knit_exit()
}

# load the mapping data
bamSummary <- bamSummarise(bamFile, blockSize=10000)

primaryBAM <- bamSummary[which(bamSummary$readFlag=="Primary"),]
unmappedBAM <- bamSummary[which(bamSummary$readFlag=="Unmapped"),]

totalSeqs <- as.numeric(validationResponse[7])
primaryMaps <- nrow(primaryBAM)
unmapped <- nrow(unmappedBAM)

totalNucleotides <- as.numeric(validationResponse[6])
mappedReadNt <- sum(primaryBAM$qwidth)
clippedReadNt <- sum(primaryBAM$coverage * primaryBAM$qwidth)  # clipped_width ...
unmappedNt <- sum(unmappedBAM$qwidth)

mapping <- data.frame(key=character(0), value=character(0), percentage=character(0))

# statistics on number of reads
mapping <- mapping %>% add_row(key="fastq reads", value=scales::comma_format()(totalSeqs), percentage="")

mapping <- mapping %>% add_row(key="mapped reads (primary)", 
                               value=scales::comma_format()(primaryMaps), 
                               percentage=paste0(round(primaryMaps / totalSeqs * 100, digits=2),"%"))
mapping <- mapping %>% add_row(key="mapped reads (secondary)", 
                               value=scales::comma_format()(length(which(bamSummary$readFlag=="Secondary"))), 
                               percentage="")
mapping <- mapping %>% add_row(key="mapped reads (supplementary)", 
                               value=scales::comma_format()(length(which(bamSummary$readFlag=="Supplementary"))), 
                               percentage="")
mapping <- mapping %>% add_row(key="unmapped reads", 
                               value=scales::comma_format()(unmapped), 
                               percentage=paste0(round(unmapped / totalSeqs * 100, digits=2),"%"))

mapping <- mapping %>% add_row(key="fastq nucleotides", value=scales::comma_format()(totalNucleotides), percentage="")

mapping <- mapping %>% add_row(key="nucleotides from primary mapped reads", 
                               value=scales::comma_format()(mappedReadNt), 
                               percentage=paste0(round(mappedReadNt / totalNucleotides * 100, digits=2),"%"))
mapping <- mapping %>% add_row(key="clipped nucleotides from primary mappings", 
                               value=scales::comma_format()(clippedReadNt), 
                               percentage=paste0(round(clippedReadNt / totalNucleotides * 100, digits=2),"%"))
mapping <- mapping %>% add_row(key="unmapped nucleotides", 
                               value=scales::comma_format()(unmappedNt), 
                               percentage=paste0(round(unmappedNt / totalNucleotides * 100, digits=2),"%"))
mapping <- mapping %>% add_row(key="nucleotides mapped (secondary mappings)", 
                               value=scales::comma_format()(sum(bamSummary[which(bamSummary$readFlag=="Secondary"),]$qwidth * bamSummary[which(bamSummary$readFlag=="Secondary"),]$coverage)), 
                               percentage="")
mapping <- mapping %>% add_row(key="nucleotides mapped (supplementary mappings)", 
                               value=scales::comma_format()(sum(bamSummary[which(bamSummary$readFlag=="Supplementary"),]$qwidth * bamSummary[which(bamSummary$readFlag=="Supplementary"),]$coverage)), 
                               percentage="")

mapping <- mapping %>% add_row(key="mean read length (primary mapping)", 
                               value=scales::comma_format()(mean(bamSummary[which(bamSummary$readFlag=="Primary"),]$qwidth)), percentage="")
mapping <- mapping %>% add_row(key="mean read length (secondary mapping)", 
                               value=scales::comma_format()(mean(bamSummary[which(bamSummary$readFlag=="Secondary"),]$qwidth)), percentage="")
mapping <- mapping %>% add_row(key="mean read length (supplementary mapping)", 
                               value=scales::comma_format()(mean(bamSummary[which(bamSummary$readFlag=="Supplementary"),]$qwidth)), percentage="")
mapping <- mapping %>% add_row(key="mean read length (unmapped)", 
                               value=scales::comma_format()(mean(bamSummary[which(bamSummary$readFlag=="Unmapped"),]$qwidth)), percentage="")


mapping <- mapping %>% add_row(key="mean read quality (primary mapping)", 
                               value=paste0(round(phredmean(bamSummary[which(bamSummary$readFlag=="Primary"),]$readq), digits=2)), percentage="")
mapping <- mapping %>% add_row(key="mean read quality (secondary mapping)", 
                               value=paste0(round(phredmean(bamSummary[which(bamSummary$readFlag=="Secondary"),]$readq), digits=2)), percentage="")
mapping <- mapping %>% add_row(key="mean read quality (supplementary mapping)", 
                               value=paste0(round(phredmean(bamSummary[which(bamSummary$readFlag=="Supplementary"),]$readq), digits=2)), percentage="")
mapping <- mapping %>% add_row(key="mean read quality (unmapped)", 
                               value=paste0(round(phredmean(bamSummary[which(bamSummary$readFlag=="Unmapped"),]$readq), digits=2)), percentage="")


mapping <- mapping %>% add_row(key="mean mapping quality (primary mapping)", 
                               value=paste0(round(mean(bamSummary[which(bamSummary$readFlag=="Primary"),]$mapq), digits=2)), percentage="")
mapping <- mapping %>% add_row(key="mean mapping quality (secondary mapping)", 
                               value=paste0(round(mean(bamSummary[which(bamSummary$readFlag=="Secondary"),]$mapq), digits=2)), percentage="")
mapping <- mapping %>% add_row(key="mean mapping quality (supplementary)", 
                               value=paste0(round(mean(bamSummary[which(bamSummary$readFlag=="Supplementary"),]$mapq), digits=2)), percentage="")


mapping <- mapping %>% add_row(key="mean coverage (primary mapping)", 
                               value=paste0(round(mean(bamSummaryToCoverage(bamFile, flag="Primary")$binned_cov), digits=2)), 
                               percentage="")
mapping <- mapping %>% add_row(key="mean coverage (secondary mapping)", 
                               value=paste0(round(mean(bamSummaryToCoverage(bamFile, flag="Secondary")$binned_cov), digits=2)), 
                               percentage="")
mapping <- mapping %>% add_row(key="mean coverage (supplementary)", 
                               value=paste0(round(mean(bamSummaryToCoverage(bamFile, flag="Supplementary")$binned_cov), digits=2)), 
                               percentage="")

  rownames(mapping) <- mapping[,1]
  mapping <- mapping[,-1]

row.names(mapping)[7]<- paste0(row.names(mapping)[7], footnote_marker_symbol(1, "html"))
row.names(mapping)[8]<- paste0(row.names(mapping)[8], footnote_marker_symbol(2, "html"))

kable(mapping, format="html", col.names=rep(" ", ncol(mapping)), caption="Table summarising mapping characteristics", booktabs=TRUE, table.envir='table*', linesep="", escape = FALSE) %>%
  kable_styling(c("striped", "condensed"))  %>%
  pack_rows("Sequence Read Mapping", 1, 5) %>%
  pack_rows("Nucleotides Mapped", 6, 11) %>%
  pack_rows("Reads lengths", 12, 15) %>%
  pack_rows("Reads Qualities", 16, 19) %>%
  pack_rows("Mapping Qualities", 20, 22) %>%
  pack_rows("Depth of Coverage", 23, 25) %>%
  footnote(symbol=c("fastq bases are calculated from the qwidth field of the primary mapped sequences", "soft-clipped nucleotide bases are calculated from the BAM CIGAR annotation"), symbol_title="please note: ", footnote_as_chunk = TRUE)
```

  
  
```{r chromosomeByChromosome, echo=FALSE}

chrIds <- getChromosomeIds()
chrIds <- chrIds[grep("(MT)|(\\.)", chrIds, invert=TRUE)]

chrMapping <- chromosomeMappingSummary(chrIds, bamFile)

kable(chrMapping, format="html", caption="Table summarising mapping characteristics", booktabs=TRUE, table.envir='table*', linesep="", escape = FALSE) %>%
  kable_styling(c("striped", "condensed"))

```
  
```{r mappingIdentity, echo=FALSE}

p1 <- plotAlignmentAccuracy(bamFile, lower=0.8) 

p2 <- plotAlignmentIdentity(bamFile, lower=0.8)

LeftRightPlot(p1, p2)

```


```{r mappingAccuracy, echo=FALSE}



plotOverallCovHistogram(bamFile) 



```
  
  
  


```{r bamSum, warning=FALSE, echo=FALSE}

Testcigar <- "2737S9M1D15M1D7M1D6M4I11M1D17M1D5M1I6M1D7M3I12M4I15M1D11M1D5M1D11M2D5M1D5M1D5M1D5M1D17M4I3M1D11M1D7M1D11M1D7M3I5M1D11M2D3M1D8M1D15M2I44M1D5M1D6M1D7M3I9M2D4M1D5M1D5M1D19M3I7M3I11M1D9M3D5M1D13M3I5M1D22M123379S"

coverageData <- bamSummaryToCoverage(bamFile, tilewidth=250000)
# filter out the non-canonical chromosomes ...
coverageData <- coverageData[-grep("(MT)|(\\.)", as.character(seqnames(coverageData))),]

plotDepthOfCoverageMegablock(coverageData)

```
  
  









```{r loadVcf, eval=FALSE}
vcf <- readVcf(file.path("results", "GM24385", "sv_calls", "GM24385_sniffles.vcf"), "hg19")
vcf <- trim(vcf)

txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

seqlevels(vcf) <- paste0("chr",seqlevels(vcf))
seqlevels(vcf, pruning.mode="coarse") <- seqlevels(vcf)[which(seqlevels(vcf) %in% seqlevels(txdb))]

locateVariants(rowRanges(vcf), txdb, AllVariants())
```


# Reproducible research - produce your own report

This report has been created using **`Rmarkdown`**, publicly available **`R`** packages, and the \LaTeX document typesetting software for reproducibility. For clarity the **`R`** packages used, and their versions, are listed below.

\fontsize{8}{12}

```{r sessionInfo, eval=TRUE, echo=FALSE, comment=NA}
utils:::print.sessionInfo(sessionInfo()[-7], locale=FALSE)
```

\fontsize{10}{14}

It is also worth recording the versions of the software that have been used for the analysis.

\fontsize{8}{12}

```{r, engine='bash', echo=FALSE, comment=NA}
conda list "samtools|minimap2|sniffles|snakemake|rstudio"
```

\fontsize{10}{14}

`r slurpContent("Static/TutorialPostamble.md")`

\pagebreak


# References and citations
